Document `KNN`

rcurtin opened the pull request:
After I don't know how many years, finally, the original thing mlpack was created for is fully documented.  I'll add links below when the documentation build finishes.

shrit reviewed (commented):
Looks very good, thank you for all the examples that you have added, I have only a couple of comments that we can fix before the merge if you would like, it is only to make things more clear

1. I would specify or add a link to explain what is the difference between exact and approximate search modes. If decided to do in our docs, then one line should be enough

![Screenshot 2025-05-30 at 12-30-13 ](https://github.com/user-attachments/assets/76b0c0c3-b96e-434d-8c97-ad4599ba5e86)


2. I would change the comments in Step 2 from "set the dataset to search in" to "Create the referenceSet for searching using the dataset". The idea here is to explain to the user when the referenceSet is created.
![Screenshot 2025-05-30 at 12-31-09 ](https://github.com/user-attachments/assets/8e3ecc7f-dfeb-43f4-8001-353a2b0fe542)

3. One thing I do not like about the SearchMode is that it is a bit confusing. So here I am thinking about the exact and the approximate search mode specified at the start of the docs, yet to discover that here we are setting the tree configurations to explore different search strategies. (I do not know if this is easily changeable but would be good to find a solution for it)
![Screenshot 2025-05-30 at 12-35-33 ](https://github.com/user-attachments/assets/9861c5ce-564a-4d3f-a04f-2973c8772854)

4. It is still not clear from the docs when the Search is Exact or approximate and how to choose one or the others.

5. If the only way to search using a referenceTree is to load a knn object model, then we should clarify that more at the start, because the way we phrased it does mean that it could be done in different ways, such as building the model in a different function as an object and use it. What I mean is, serializtion is only mention at the end in the examples.

rcurtin commented:
@shrit I think I've addressed your concerns with a handful of fairly major changes.

 * I changed the `NeighborSearchMode` enum to `NeighborSearchStrategy`, and changed all names like `DUAL_TREE_MODE` to `DUAL_TREE`.  I had to keep the old enum for reverse compatibility; we can remove it all once mlpack 5.0.0 comes around.  I updated all the documentation throughout, so now the word "strategy" is used to refer to whether we are doing dual-tree/single-tree/brute force/greedy single-tree.
 
 * To clarify the difference between the reference and query sets, I added an SVG diagram to the top of the page showing some simple examples, including what is meant by approximation.

 * Instead of awkward phrasing about 'the set to be searched in', I introduced the term 'reference set' in the beginning of the documentation and pretty much stick with it throughout.  There are still a couple places where I kept the original phrasing, but only when it is very clear that I am talking about the reference set anyway.
 
Hopefully this addresses the concerns you had... take a look once the documentation builds and let me know what you think.

shrit commented on src/mlpack/methods/neighbor_search/kfn_main.cpp:
```diff
@@ -179,16 +179,16 @@ void BINDING_FUNCTION(util::Params& params, util::Timers& timers)
 
   if (algorithm == "naive")
-    searchMode = NAIVE_MODE;
+    searchStrategy = NAIVE;
   else if (algorithm == "single_tree")
-    searchMode = SINGLE_TREE_MODE;
+    searchStrategy = SINGLE_TREE;
   else if (algorithm == "dual_tree")
-    searchMode = DUAL_TREE_MODE;
+    searchStrategy = DUAL_TREE;
   else if (algorithm == "greedy")
-    searchMode = GREEDY_SINGLE_TREE_MODE;
+    searchStrategy = GREEDY_SINGLE_TREE;
 
```
:+1:

shrit reviewed (approved):
Perfect, this looks good. thank you for the modification. I feel like this small suggestion made you modify half of the library.
Agreed, all points have been addressed, let us get this PR merged.

shrit commented:
@mlpack-jenkins test this please
